File: frontend/src/features/activity/activityStatus.ts
export type StatusFlags = {
  offline: boolean
  pending: boolean
  busy: boolean
  hasSnapshot: boolean
  error: string | null
}

export function compactStatusLabel(flags: StatusFlags) {
  if (!flags.hasSnapshot) return 'Connection required'
  if (flags.offline && flags.busy) return 'Offline · Saving…'
  if (flags.offline && flags.pending) return 'Offline · Unsynced'
  if (flags.offline) return 'Offline'
  if (flags.busy) return 'Saving…'
  if (flags.pending || flags.error) return 'Unsynced'
  return 'Synced'
}

export function statusAttention(flags: StatusFlags) {
  return flags.offline || flags.pending || flags.busy || !flags.hasSnapshot || flags.error != null
}

export function statusMarkKind(flags: StatusFlags) {
  return statusAttention(flags) ? 'chip' : null
}

export function statusShowsRetry(flags: StatusFlags) {
  return flags.error != null || flags.pending
}

File: android/app/src/main/java/com/timebox/android/ui/day/ActivityStatus.kt
package com.timebox.android.ui.day

data class StatusFlags(
    val offline: Boolean,
    val pending: Boolean,
    val busy: Boolean,
    val hasSnapshot: Boolean,
    val error: String?,
)

fun compactStatusLabel(flags: StatusFlags) = when {
    !flags.hasSnapshot -> "Connection required"
    flags.offline && flags.busy -> "Offline · Saving…"
    flags.offline && flags.pending -> "Offline · Unsynced"
    flags.offline -> "Offline"
    flags.busy -> "Saving…"
    flags.pending || flags.error != null -> "Unsynced"
    else -> "Synced"
}

fun statusAttention(flags: StatusFlags) =
    flags.offline || flags.pending || flags.busy || !flags.hasSnapshot || flags.error != null

fun statusMarkKind(flags: StatusFlags) = if (statusAttention(flags)) "chip" else null

fun statusShowsRetry(flags: StatusFlags) = flags.error != null || flags.pending
